### PR TITLE
Fix applying `SPECIFIC_PRODUCT` and `apply_once_per_order` voucher on draft orders

### PR DESCRIPTION
**Pull request by IKarbowiak**
Apply `SPECIFIC_PRODUCT` and `apply_once_per_order` voucher on draft orders.

Fixes: https://github.com/saleor/saleor/issues/15517

Internal issues:
- https://linear.app/saleor/issue/SHOPX-873/voucher-with-apply-once-per-order-not-working-properly
- https://linear.app/saleor/issue/SHOPX-872/specific-products-vouchers-in-draft-order-do-not-work

# Impact

- [ ] New migrations
- [ ] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [ ] Link to documentation:

# Pull Request Checklist

- [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
- [ ] Database queries are optimized and the number of queries is constant
- [ ] Database migrations are either absent or optimized for zero downtime
- [ ] The changes are covered by test cases
- [ ] All new fields/inputs/mutations have proper labels added (`ADDED_IN_X`, `PREVIEW_FEATURE`, etc.)
- [ ] All migrations have proper dependencies
- [ ] All indexes are added concurrently in migrations
- [ ] All RunSql and RunPython migrations have revert option defined

### CODE REVIEW

**Comment by andrzejewsky** [comment]
Oh! That's a wonderful pull request! Fantastic work! 🥳

**Review by Air-t — APPROVED** [review]

File: saleor/checkout/fetch.py [hunk 1]
```diff
@@ -34,44 +35,23 @@
...
         ProductVariantChannelListing,
     )
     from ..tax.models import TaxClass, TaxConfiguration
-    from .models import Checkout, CheckoutLine
+    from .models import Checkout
 
 
 @dataclass
-class CheckoutLineInfo:
+class CheckoutLineInfo(LineInfo):
```
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/checkout/fetch.py:49]: It's checkout specific so I decided to left it in `checkout` module, we can discuss this with the team

File: saleor/core/pricing/interface.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    ):
+        return [
+            discount
+            for discount in self.discounts
+            if discount.type in [DiscountType.PROMOTION, DiscountType.ORDER_PROMOTION]
+        ]
+
+    def get_catalogue_discounts(
+        self,
+    ):
```
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/core/pricing/interface.py:45]: Same as above

File: saleor/discount/utils.py [hunk 3]
```diff
@@ -186,6 +187,81 @@ def get_voucher_code_instance(
...
+        or voucher_info.collection_pks
+        or voucher_info.category_pks
+        or voucher_info.variant_pks
+    ):
+        for line_info in lines:
+            if line_info.line.is_gift:
+                continue
+            line_variant = line_info.variant
+            line_product = line_info.product
+            if not line_variant or not line_product:
```
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/discount/utils.py:234]: No, but without those checks mypy is crying

[the remark this answers is not in the document]
Comment (korycins) [reply saleor/discount/utils.py:234]: `LineInfo` has `Optional[Product]`.  For `CheckoutLineInfo` we modify it to `"Product"`, but we don't do that for `DraftOrderLineInfo`.  So it is expected that mypy complain here, as based on `LineInfo`, `Product` can be None. 
Isn't like incorrect typing for `DraftOrderLineInfo`? Shouldn't we also require to have a `Product` there?

File: saleor/discount/utils.py [hunk 4]
```diff
@@ -186,6 +187,81 @@ def get_voucher_code_instance(
...
+def apply_voucher_to_line(
+    voucher_info: "VoucherInfo",
+    lines_info: Iterable["LineInfo"],
+):
+    """Attach voucher to valid checkout or order lines info.
+
+    Apply a voucher to checkout/order line info when the voucher has the type
+    SPECIFIC_PRODUCTS or is applied only to the cheapest item.
+    """
+    from .utils import get_discounted_lines
```
Comment (korycins) [inline saleor/discount/utils.py:199]: Do you still need this? `get_discounted_lines` is placed below this function.

File: saleor/discount/utils.py [hunk 5]
```diff
@@ -186,6 +187,81 @@ def get_voucher_code_instance(
...
+        cheapest_line = _get_the_cheapest_line(lines_included_in_discount)
+        if cheapest_line:
+            discounted_lines_by_voucher = [cheapest_line]
+    for line_info in lines_info:
+        if line_info in discounted_lines_by_voucher:
+            line_info.voucher = voucher
+            line_info.voucher_code = voucher_info.voucher_code
+
+
+def get_discounted_lines(
```
Comment (korycins) [inline saleor/discount/utils.py:219]: Correct me if I am wrong. `get_discounted_lines` is a replacement for custom `get_discounted_lines` functions used with Checkout, and Order (`saleor.checkout.utils.get_discounted_lines`, `saleor.order.utils.get_discounted_lines`). Shouldn't we fully drop these functions?

File: saleor/order/fetch.py [hunk 6]
```diff
@@ -114,6 +109,7 @@ def fetch_draft_order_lines_info(
     channel = order.channel
     for line in lines:
         variant = cast(ProductVariant, line.variant)
+        product = variant.product if variant else None
```
Comment (korycins) [inline saleor/order/fetch.py:112]: You have `cast` and `if variant else None`. So which one is correct here? Variant can be null or variant is always provided?

File: saleor/order/fetch.py [hunk 7]
```diff
@@ -127,12 +123,22 @@ def fetch_draft_order_lines_info(
...
                 channel_listing=variant_channel_listing,
                 discounts=list(line.discounts.all()),
                 rules_info=rules_info,
                 channel=channel,
+                voucher=None,
+                voucher_code=None,
             )
         )
+    voucher = order.voucher
+    if voucher and (
```
Comment (korycins) [inline saleor/order/fetch.py:137]: In `saleor.order.calculations.fetch_order_prices_if_expired` we also call `saleor.order.calculations._update_order_discount_for_voucher`. It creates an orderDiscount object. I don't see any condition for specfic_product or apply_once_per_order. It means, that we will propagate the discount over lines and create the additional order.discounts objects. Is it expected? Won't we add a discount twice?

Comment (IKarbowiak) [reply saleor/order/fetch.py:137]: It's not the correct behaviour, we have a ticket for fixing that. I want to narrow the changes here as it's quite big right now. The `OrderDiscount` object will have `0` discount amount, so the discount won't be apply twice, but still you are right that the object shouldn't be created. We will fix that in another PR.

File: saleor/discount/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/discount/utils.py:234]: You are right. All draft order lines will have `variant` instance, so both `ProductVariant` and `Product` cannot be `None`

File: saleor/discount/utils.py (hunk 4 above)
Comment (IKarbowiak) [reply saleor/discount/utils.py]: Fixed

File: saleor/order/fetch.py (hunk 6 above)
Comment (IKarbowiak) [reply saleor/order/fetch.py]: fixed

File: saleor/discount/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/discount/utils.py:232]: @korycins fixed

**Comment by tomaszszymanski129** [comment]
@IKarbowiak note some tests are failing with `AttributeError`

File: saleor/discount/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/discount/utils.py]: I double checked and both `variant` and `product` must be optional on `DraftOrderLineInfo`, as there might be the possibility that `variant` for unconfirmed order has been removed.

**Review by korycins — APPROVED** [review]

**Review by szczecha — APPROVED** [review]

**Pull request merged**